design/secret_rotation_design.md

```diff
@@ -0,0 +1,951 @@
+| All Kubernetes Secrets files are overwritten if custom template is changed in Conjur | GA | `k8s-secret` | |
+| Convert to signal number if `restart-app-signal` is configured with valid signal name | Future | Either | |
+| Exit with error if `restart-app-signal` is configured with invalid signal name | Future | Either | |
+| Processes other than Secret Provider's process and `/pause` process are sent restart signal when Secrets/file content is updated. | Future | Either | |
+| Secret Provider does not send restart signal to its own process when Secrets/file content is updated. | Future | Either | |
+| Secret Provider does not send restart signal to `/pause` process when Secrets/file content is updated. | Future | Either | |
+
+#### E2E tests
+
+E2E tests are expensive. We should limit E2E testing to happy path "smoke
+tests", and leave sad path testing to unit testing and lower level functional
+tests.
+
+There already exists several E2E test scenarios that test Secrets Provider
+functionality in init container and standalone, application container mode,
+for both Kubernetes Secrets destination mode and secret file mode.
+
+The introduction of the secret rotation feature will require the addition
+of a few happy path E2E test scenarios:
+
+- [ ] Happy path, Secrets Provider running as a sidecar without secrets rotation
+- [ ] Happy path, Secrets Provider running as a sidecar with secrets rotation
+- [ ] Secret files get restored following app crash with Secrets Provider running as sidecar
+
+#### Security testing
+
+Security testing will include:
+
+- Automated Vulnerability scans
+
+####  Performance testing
+
+Performance of the secrets rotation feature will be measured/confirmed
+**manually** by configuring the Secrets Provider with a typical,
+representative configuration in terms of the number of secret groups and the
+number of secrets per secret group:
+
+- 5 secret groups
+- 5 secrets / secret group
+- Mix of secret value lengths (10 to 100 characters, plus some base64-encoded
+  TLS certificates)
+
+This profile represents about half of the maximum 50 Conjur secrets per
+Secrets Provider container scaling limit.
+
+Latency measurements will be taken by using the Secrets Provider
+OpenTelemetry tracing feature.
+
+## Logs
+
+For "happy path" operation, the secrets rotation feature will leverage INFO
+level logs that are already built into the Secrets Provider to report status
+for significant checkpoints in the workflow:
+
+- Start of annotation parsing
+- Completion of annotation parsing
+- Start of content rendering for each Kubernetes Secret / secrets file
+- Start of each Secrets update / write of secrets file
+- Completion of each Secrets update / write of secrets file
+
+For error conditions, the Secrets Provider will include ERROR-level logs.
+Annotation values are treated as non-sensitive and will not be redacted
+in log messages.
```

Comment:
Should check that if natefinch/atomic fails in writing out the file, that it doesn't echo what would have been the file contents in its error message.  Also related, but not logs, it looks like atomic tried to remove the temp file it uses if an error occurs, which is good, but simply ignores any error in removing that temp file. Do we need to have a final check to make sure we didn't leave a temp file with secrets values somewhere the k8s admin doesn't expect?

Reply:
> it looks like atomic tried to remove the temp file it uses if an error occurs, which is good, but simply ignores any error in removing that temp file. Do we need to have a final check to make sure we didn't leave a temp file with secrets values somewhere the k8s admin doesn't expect?

We can, or we can fork the project and change the behavior so it doesn't ignore the error. What would we do if we found that a temp file couldn't be deleted? Just print a message to the log?

Reply:
I think we'd need to log it at least. I don't know if we'd have escalation paths built in (sort of like, "if kill doesn't work, try kill -9") to see if we could remove it on subsequent tries, but those might be possibilities to consider as well. I'm guessing not since it's atomic not handling it. Not sure if there's a time component to this either, but a simple retry maybe? Again probably not an issue most of the time and in the rare cases it might be, not clear there's much we can do. We should definitely dig in and make sure we do anything we can not to leave secret values hanging around in a temp file somewhere though.

Reply:
Here's a draft that implements what I understand you to be saying.
PR https://github.com/cyberark/secrets-provider-for-k8s/pull/440, and specifically commit [9e0f689](https://github.com/cyberark/secrets-provider-for-k8s/pull/440/commits/9e0f6894a7fa30afaa914acb844c9aa1c1cf78dc) (this uses a forked version of the 'atomic' library to report whether the temp file was deleted)

Reply:
I think that looks good. Thank you! Different security hat - with a forked version, do we have mechanisms in place to keep our fork uptodate as the base package evolves?